fix(core): remove actix dependency from near-primitives

fckt opened the pull request:
The PR follows the very specific goal - to remove `actix` dependency from near-primitives.

The general approach here is  to move in small iterations to bring us closer to https://github.com/nearprotocol/nearcore/issues/2433

nearmax commented on env/Cargo.toml:
```diff
@@ -0,0 +1,9 @@
+[package]
```
This crate is essentially a place where we dump stuff that we don't know where else to put. I don't think root of nearcore is a good place for it and I think it should have a more descriptive name.

In general, instead of extracting actix from the primitives crate we should be extracting stuff used by the runtime into finer grained crates, leaving everything else behind in the primitives crate. It requires revisiting every type from primitives used by the runtime and deciding which category of primitives it better represents. It is a much harder thing to do than just cutting out actix from dependencies, but it actually what we want to do -- we want to break monolithic primitives crate into smaller but meaningful crates.

frol commented:
I believe that the best documentation is a self-descriptive source code. I feel that there is some room for improvement in nearcore in this direction, so whatever opportunity we have with refactoring, we should take it to improve things.

fckt commented:
Thank you for your feedback, folks.

> I agree with Max, this new crate works around the dependency issue, but it has a completely non-descriptive name. Given we cannot come up with a better name, I consider it is an indicator that we need to think more about the proper separation.

You're right, this is hard to come up with a good name for this (but it's better the to export this from near-primitives). The goal of this change is not bring a perfect solution.

What's makes up this crate unique:
- by type: it's a library
- by functionality: it accumulates functions which alters globals
- by usage: used only by binaries (main + tests) and not libraries

Keeping in mind these properties I don't see a category which suits all properties listed (in https://github.com/nearprotocol/nearcore/issues/1881#issue-540289254):

Maybe the best approach would be just having in in a `neard` category, but expose it as lib with name `neard-env` which would clearly symbolize that it's just a collection of tools which alters env in which neard executes. WDYT?

frol commented:
We should never consider making library-crate dependent on neard.

`env` and `neard-env` does not help me to make any guess about the type of functionality this crate may expose. `near-actix-helpers`, `near-tracing-helpers`, `near-testlib-helpers` (though, if we use those helpers outside of tests, it is not a good name), ...

I think, `libraries/helpers/` is a good candidate to score such crates

nearmax commented:
Overall, the goal is to be able to publish runtime and compile it to Wasm. We should not be extracting stuff from primitives until it becomes publishable. Our crates should be as semantic as possible, having a crate that represents "stuff that does not compile to Wasm" is not good. Instead we should be extracting crates from `primitives` into finer-grained semantic crates. E.g. the way we extracted `crypto` and `runtime-configs`.  Runtime should not be linked against something it does not care about, like `network.rs` or `sharding.rs`.

However, that said, this is probably a huge effort which would require too much refactoring that would delay our progress with testing harness, so let's create an issue for it right now, and link runtime against the primitives crate (with near-actix-helpers, near-tracing-helpers, near-testlib-helpers extracted).

> I think, libraries/helpers/ is a good candidate to score such crates

I suggest put it in `utils` folder, almost every crate in our infra is a library, so no reason to have `libraries` folder. We also have `test-utils` already.

frol commented:
@nearmax In #1881, I analyzed the root folder of nearcore project. There are tons of files and folders, so I categorized the folders into:

* "neard" (we need to have the main entry-point to the project visible)
* "tools" (near-loadtester, near-state-viewer, near-genesis)
* "libraries" (which has subcategories: "core", "chain", "runtime", "test-utils", and "helpers" [though "utils" is also fine])

We have a ton of other files in the root of nearcore, which we cannot move, so it makes sense to shrink the scope where we can, so I think "libraries" folder is really helpful.

> However, that said, this is probably a huge effort which would require too much refactoring that would delay our progress with testing harness

Agree, it is the tradeoff we want to take now to reach the goal. I think even this straightforward refactoring of extracting obviously non-related parts into separate crates will create a precedent, and make the future development cleaner.

fckt commented:
IMO, we should have the complete components (categorized by their role in the system) on the top level, but not by the type of their interface.

fckt commented:
Addressed concerns 
https://github.com/nearprotocol/nearcore/pull/2511/commits/fdf96336a024e2b87f744a9c9de52ac192f11f11
@nearmax @frol @bowenwang1996

fckt commented on Cargo.toml:
```diff
@@ -12,6 +12,7 @@ members = [
     "core/runtime-configs",
     "core/store",
     "core/metrics",
+    "utils/actix",
```
I agree with you, but the entire project doesn't follow this for now. As long there is no such a convention, I don't see too much reason to we enforce it selectively.

frol replied:
Crates in `runtime` folder partially follow the rule, and I think we can pull out this renaming of all the folders in place as one of the steps towards greater refactoring.

fckt replied:
I agree. I'd suggest to create a separate issue/PR for it since it's clearly out of scope of this PR.

frol replied:
We don't want to rename the rest of the crates in the scope of this PR, but let's rename those you introduce here.

fckt replied:
There are some pro and const in both variants. IMO the consistency with the currently dominated style is more important. In the same time, I really don't think it's a critical decision to make.

fckt commented on Cargo.toml:
```diff
@@ -47,6 +48,7 @@ serde_json = "1.0.0"
 reqwest = "0.10"
 futures = "0.3"
 
+near-logger-utils = { path = "./test-utils/logger" }
```
ditto

nearmax commented on utils/actix/Cargo.toml:
```diff
@@ -0,0 +1,8 @@
+[package]
```
Is there a specific reason why crate name is different from the folder name? Same for the logger.

fckt replied:
It follows the current de-facto convention throughout the project. Discussed here: https://github.com/nearprotocol/nearcore/pull/2511#pullrequestreview-399223765

evgenykuzyakov commented on core/primitives/src/account.rs:
```diff
@@ -20,6 +20,17 @@ pub struct Account {
+            amount,
+            locked: 0,
+            code_hash,
+            storage_usage: std::mem::size_of::<Account>() as u64,
```
This is not the default value we use. We use `100` storage usage right now for a new account, so we can't switch to mem size of the account structure.  This is the reason it was in the `test_utils`

frol replied:
@evgenykuzyakov Good catch!

@fckt What is the reason we moved all the stuff from `test_utils.rs`? It is not great to have it there, but neither it is good to have those methods mixed with the production methods.

I vote for having those test methods hidden behind an ad-hoc `Testing` trait (use `easy-ext` for easy creation of the ad-hoc traits). In any case, we should not have these methods mixed with the production implementations.

fckt replied:
> @fckt I vote for having those test methods hidden behind an ad-hoc Testing trait (use easy-ext for easy creation of the ad-hoc traits). In any case, we should not have these methods mixed with the production implementations.

This was exactly a reason. But I didn't know that this cannot be used as a default constructor.

> We use 100 storage usage right now for a new account, so we can't switch to mem size of the account structure.

thanks, good to know

fckt replied:
https://github.com/nearprotocol/nearcore/pull/2511/#issuecomment-620906318

frol commented on core/primitives/src/transaction.rs:
```diff
@@ -192,6 +197,89 @@ impl SignedTransaction {
+        block_hash: CryptoHash,
+    ) -> Self {
+        Self::from_actions(
+            nonce,
+            originator,
+            new_account_id,
+            signer,
+            vec![
+                Action::CreateAccount(CreateAccountAction {}),
+                Action::AddKey(AddKeyAction {
+                    public_key,
+                    access_key: AccessKey { nonce: 0, permission: AccessKeyPermission::FullAccess },
+                }),
+                Action::Transfer(TransferAction { deposit: amount }),
+            ],
+            block_hash,
+        )
+    }
+
+    pub fn delete_account(
+        nonce: Nonce,
+        signer_id: AccountId,
+        receiver_id: AccountId,
+        beneficiary_id: AccountId,
+        signer: &dyn Signer,
+        block_hash: CryptoHash,
+    ) -> Self {
+        Self::from_actions(
+            nonce,
+            signer_id,
+            receiver_id,
+            signer,
+            vec![Action::DeleteAccount(DeleteAccountAction { beneficiary_id })],
+            block_hash,
+        )
+    }
+
+    pub fn empty(block_hash: CryptoHash) -> Self {
+        Self::from_actions(0, "".to_string(), "".to_string(), &EmptySigner {}, vec![], block_hash)
+    }
```
These test methods also pollute the public API. The best is to move these methods to testlib, and also introduce the SignedTransactionTesting trait (easy-ext does this effortless).

Apply the same thinking to the rest methods you moved out of `test_utils.rs`

fckt replied:
see https://github.com/nearprotocol/nearcore/pull/2511#issuecomment-620906318

frol commented on test-utils/testlib/src/runtime_utils.rs:
```diff
@@ -8,12 +8,16 @@ use near_chain_configs::Genesis;
 use neard::config::GenesisExt;
 use node_runtime::{state_viewer::TrieViewer, Runtime};
 
+pub fn account_new(amount: Balance, code_hash: CryptoHash) -> Account {
```
`init_test_account` reads better for me.

fckt replied:
init_* is already occupied by methods which are alter global static vars

fckt commented:
I realized that having `test_utils` module in `near-*` crates is a common practice for today. I restored the `core/primitives/src/test_utils.rs`.
